// weave/dist/opType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpRef = void 0;
exports.isOp = isOp;
exports.getOpWrappedFunction = getOpWrappedFunction;
exports.getOpName = getOpName;
exports.getOpParameterNames = getOpParameterNames;
const urls_1 = require("./urls");
function isOp(value) {
    return value && value.__isOp === true;
}
function getOpWrappedFunction(opValue) {
    return opValue.__wrappedFunction;
}
function getOpName(opValue) {
    return opValue.__name;
}
function getOpParameterNames(opValue) {
    return opValue.__parameterNames;
}
class OpRef {
    constructor(projectId, objectId, digest) {
        this.projectId = projectId;
        this.objectId = objectId;
        this.digest = digest;
    }
    // TODO: Add extra
    uri() {
        return `weave:///${this.projectId}/op/${this.objectId}:${this.digest}`;
    }
    ui_url() {
        const domain = (0, urls_1.getGlobalDomain)();
        return `https://${domain}/${this.projectId}/weave/ops/${this.objectId}/versions/${this.digest}`;
    }
    async get() {
        throw new Error('Not implemented');
    }
}
exports.OpRef = OpRef;
