// weave/dist/prompt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessagesPrompt = exports.StringPrompt = exports.Prompt = void 0;
const weaveObject_1 = require("./weaveObject");
class Prompt extends weaveObject_1.WeaveObject {
    constructor(parameters) {
        super(parameters);
    }
}
exports.Prompt = Prompt;
class StringPrompt extends Prompt {
    constructor(parameters) {
        super(parameters);
        this.content = parameters.content;
    }
    format(values = {}) {
        return formatString(this.content, values);
    }
}
exports.StringPrompt = StringPrompt;
class MessagesPrompt extends Prompt {
    constructor(parameters) {
        super(parameters);
        this.messages = parameters.messages;
    }
    formatMessage(message, values = {}) {
        const formattedMessage = Object.fromEntries(Object.entries(message).map(([key, value]) => {
            if (typeof value === 'string') {
                return [key, formatString(value, values)];
            }
            else if (Array.isArray(value) &&
                value.every(item => typeof item === 'object' && item !== null)) {
                return [key, value.map(item => this.formatMessage(item, values))];
            }
            else {
                return [key, value];
            }
        }));
        return formattedMessage;
    }
    format(values = {}) {
        return this.messages.map(message => {
            return this.formatMessage(message, values);
        });
    }
}
exports.MessagesPrompt = MessagesPrompt;
function formatString(str, kwargs = {}) {
    return str.replace(/\{(\w+)(:[^}]+)?\}/g, (match, key) => {
        return kwargs[key] !== undefined ? kwargs[key] : match;
    });
}
