// weave/dist/table.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Table = exports.TableRowRef = exports.TableRef = void 0;
class TableRef {
    constructor(projectId, digest) {
        this.projectId = projectId;
        this.digest = digest;
    }
    uri() {
        return `weave:///${this.projectId}/table/${this.digest}`;
    }
}
exports.TableRef = TableRef;
class TableRowRef {
    constructor(projectId, digest, rowDigest) {
        this.projectId = projectId;
        this.digest = digest;
        this.rowDigest = rowDigest;
    }
    uri() {
        return `weave:///${this.projectId}/table/${this.digest}/id/${this.rowDigest}`;
    }
}
exports.TableRowRef = TableRowRef;
class Table {
    constructor(rows) {
        this.rows = rows;
    }
    get length() {
        return this.rows.length;
    }
    async *[Symbol.asyncIterator]() {
        for (let i = 0; i < this.length; i++) {
            yield this.row(i);
        }
    }
    row(index) {
        return this.rows[index];
    }
}
exports.Table = Table;
