// weave/dist/urls.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultDomain = exports.defaultHost = void 0;
exports.getUrls = getUrls;
exports.getGlobalDomain = getGlobalDomain;
exports.setGlobalDomain = setGlobalDomain;
exports.defaultHost = 'api.wandb.ai';
exports.defaultDomain = 'wandb.ai';
function getUrls(host) {
    const resolvedHost = host !== null && host !== void 0 ? host : exports.defaultHost;
    const isDefault = resolvedHost === exports.defaultHost;
    let traceBaseUrl = process.env.WF_TRACE_SERVER_URL;
    if (traceBaseUrl === undefined || traceBaseUrl === null) {
        traceBaseUrl = isDefault
            ? `https://trace.wandb.ai`
            : `https://${resolvedHost}/traces`;
    }
    return {
        baseUrl: isDefault ? `https://api.wandb.ai` : `https://${resolvedHost}`,
        traceBaseUrl,
        domain: isDefault ? exports.defaultDomain : resolvedHost,
        host: isDefault ? exports.defaultHost : resolvedHost,
    };
}
let globalDomain = undefined;
function getGlobalDomain() {
    return globalDomain;
}
function setGlobalDomain(domain) {
    globalDomain = domain;
}
